### PR TITLE
Add Vehicle Exit Point functionality to E2 with extension (Fixes #1373)

### LINKED ISSUE #1373: E2: Add Vehicle Exit Point functionality
Simple. Add functionality of Vehicle Exit Point to Expression 2:
`entity:getVehicleExitAng()`: Gets exit angles of a vehicle (this will return exit-point angle only if you have set it via setVehicleExitAng; otherwise, it will return invalid/empty angle).
`entity:getVehicleExitPos()`: Gets exit position of a vehicle (this will return exit-point position only if you have set it via setVehicleExitPos; otherwise, it will return invalid/empty vector).
`entity:removeVehicleExitAng()`: Removes exit angles of a vehicle.
`entity:removeVehicleExitPos()`: Removes exit position of a vehicle.
`entity:setVehicleExitAng(angle)`: Sets exit angles of a vehicle.
`entity:setVehicleExitPos(vector)`: Sets exit position of a vehicle.
And also make it respect `wire_pod_exit_distance` ConVar value (the maximum distance an exit-point can move a player, by default it is set to 1000 units).

**I am working on this**.

React to this issue:
Vote 👍 to support; otherwise 👎

Consider independent E2s. I support :)

_edit_:
I have pushed them, here they are, two pull requests as promised:
[Without E2 extension](https://github.com/wiremod/wire/pull/1378)
[With E2 extension](https://github.com/wiremod/wire/pull/1379)

Now it is Your turn, go vote again people, this time you **vote for the one to be merged** in official Wire :)
The one with most up-votes (👍) should be merged (you can't vote for both!).
**Please, also post your reason after voting; otherwise, your vote won't count!
For example: Why did you vote for it instead of voting for the other one?**

### PR DESCRIPTION
**Pull request by CaptainPRICE**
Fixes #1373.

This pull request (after merged) would add the following to Expression 2 gate:

new Functions:

| Signature | Description | Cost |
| --- | --- | --- |
| `angle=entity:getVehicleExitAng()` | Gets exit angles of a vehicle (this will return exit-point angle only if you have set it via setVehicleExitAng; otherwise, it will return invalid/zero angle) | 5 |
| `vector=entity:getVehicleExitPos()` | Gets exit position of a vehicle (this will return exit-point position only if you have set it via setVehicleExitPos; otherwise, it will return invalid/origin vector) | 5 |
| `entity:removeVehicleExitAng()` | Removes exit angles of a vehicle (this will remove exit-point angle only if you have set it via setVehicleExitAng) | 5 |
| `entity:removeVehicleExitPos()` | Removes exit position of a vehicle (this will remove exit-point position only if you have set it via setVehicleExitPos) | 5 |
| `entity:setVehicleExitAng(angle)` | Sets exit angles of a vehicle | 10 |
| `number=entity:setVehicleExitPos(vector)` | Sets exit position of a vehicle (returns 1 on success; otherwise, it returns 0 to indicate a failure) | 10 |

**This is edition which DOES implement it via Expression 2 extension**, I have called it `vehicle`.

I have allowed edits from maintainers, feel free to get rid of the comments from the code if you wish...
I have tested it in-game - it works as expected, and it co-operates with `wire_pod_exit_distance` ConVar (the maximum distance an exit-point can move a player).

[With E2 extension](https://github.com/wiremod/wire/pull/1379) vs [Without E2 extension](https://github.com/wiremod/wire/pull/1378)
Now it is Your turn to vote, this time you **vote for the one to be merged** in official Wire :)
The one with most up-votes (👍) should be merged (you can't vote for both!).
**Please, also post your reason after voting; otherwise, your vote won't count!
For example: Why did you vote for THIS one?**

### CODE REVIEW

File: lua/entities/gmod_wire_expression2/core/vehicle.lua [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+--local util_TraceHull = util.TraceHull
+local util_PointContents = util.PointContents
+
+registerCallback("e2lib_replace_function", function(funcname, func, oldfunc)
+	if funcname == "IsValid" then
+		IsValid = func
+	elseif funcname == "isOwner" then
+		isOwner = func
+	end
+end)
```
Comment (Nebual) [inline lua/entities/gmod_wire_expression2/core/vehicle.lua]: I think you should just call E2Lib.isOwner directly without bothering to localize it, and I don't think IsValid is ever replaced - its a Garry global I believe.

In general localization like this is very much a micro-optimization anyway.

File: lua/entities/gmod_wire_expression2/core/vehicle.lua [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	hook.Add("EntityRemoved", "E2RemoveVehicleExitPoint", function(ent)
+		-- No need for redundant checks, if <ent> key exists in VehicleExitPoints table, data will be removed by setting it to nil value.
+		VehicleExitPoints[ent] = nil
+	end)
+end)
+
+--registerCallback("construct", function(...) print("construct", ...) end)
+
+-- TODO: Possibly remove exit-point data from vehicles assigned by E2 chip (when it is removed)?
+--registerCallback("destruct", function(...) print("destruct", ...) end)
```
Comment (Nebual) [inline lua/entities/gmod_wire_expression2/core/vehicle.lua]: I think I'd prefer if the E2 destructor cleaned up the changes it made to vehicles.

File: lua/entities/gmod_wire_expression2/core/vehicle.lua [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	elseif funcname == "isOwner" then
+		isOwner = func
+	end
+end)
+
+--------------------------------------------------------------------------------
+
+local wire_pod_exit_distance = GetConVar("wire_pod_exit_distance") -- ConVar is created in Vehicle Exit Point entity. We use this to get the maximum distance an exit-point can move a player (by default it is set to 1000 units).
+local VehicleExitPoints = {} -- This will contain the exit-point data (indexed by Vehicle).
+local InvalidExitValue = { 0, 0, 0 } -- Localized table for "invalid" angle/vector, heck why not, it is better this way than creating a table each time.
```
Comment (Nebual) [inline lua/entities/gmod_wire_expression2/core/vehicle.lua]: I'd prefer if almost all of the comments in this file were removed - some of them are obviously your unused code, and most of the function descriptions are pretty self-evident from the function signatures themselves.

The only one's I'd suggest keeping are these 3 variable explanations, L106's trace explanation, I guess L122 is ok.

**Comment by CaptainPRICE** [comment]
Personally I preferred this version.
It is gone. I do still have a backup of the repository on my old hard drive tho.. I don't know if I have enough motivation to ditch this. RIP...

**Pull request closed without merging**